hybrid/flow.py

```diff
@@ -745,51 +754,58 @@ def iteration_update(self, iterno, cnt, input_state, output_state):
 
     def next(self, state):
         iterno = 0
-        cnt = self.convergence
+        cnt = self.convergence or 0
         input_state = state
+        output_state = input_state
 
-        while True:
+        while not self._stop_event.is_set():
             output_state = self.runnable.run(input_state, executor=immediate_executor).result()
 
             iterno, cnt, input_state = self.iteration_update(iterno, cnt, input_state, output_state)
 
             if self.max_iter is not None and iterno >= self.max_iter:
                 break
-            if cnt <= 0:
+            if self.convergence is not None and cnt <= 0:
                 break
 
         return output_state
 
-
-class LoopN(LoopUntilNoImprovement):
-    """Iterate `runnable` exactly `n` times, using each output as input in the
-    next iteration.
-    """
-
-    def __init__(self, runnable, n):
-        super(LoopN, self).__init__(
-            runnable=runnable, max_iter=n, convergence=n, key=lambda _: 0)
+    def halt(self):
+        self.runnable.stop()
+        self._stop_event.set()
 
 
 class Loop(LoopUntilNoImprovement):
     pass
 
+
 SimpleIterator = Loop
 
 
 class LoopWhileNoImprovement(LoopUntilNoImprovement):
     """Iterates `runnable` until a state quality metric, defined by the `key`
     function, shows no improvement for at least `max_tries` number of
-    iterations.
+    iterations (or until `max_iter` number of iterations is exceeded).
+
+    Note:
+        Unlike `LoopUntilNoImprovement`/`Loop`, `LoopWhileNoImprovement` will
+        run the loop body runnable with the same input if it shows no
+        improvement, and it will use the new output if it's better than previous.
```

Comment:
Unclear:
* "loop body" --> why not the standard "iterate 'Runnable'" talk?
* "will run ... if it" --> is the running really conditional upon no improvement or do you mean that it iterates regardless of any improvement?
* "if it shows no improvement? --> "it" refers to the input. Is it not looking for the improvement on the output?

Reply:
1) I used "loop body runnable" as an analogy with a standard loop (which has a body that's iterated over)
2) the emphasis should have been on the "same input", I rephrased it
3) fixed